### PR TITLE
Add `Project.add` helper

### PR DESCRIPTION
**Pull request by alexdavid**

### CODE REVIEW

**Comment by soenkehahn** [comment]
Trying this out in both #393 and #397 it seems that this solves our use-cases there. I think there's two open questions:

- Is `.add` the best name? Some alternatives are `.with`, `.modify`, `.withSelf`. Any other ideas? I kind of like both `.add` and `.with`. Although, technically, this could be used to *not* add something but somehow modify the project in another way, including *removing* things. So maybe `.add` is not the best name.
- Is there a good way to provide a type alias for these plugins? Something like `type Plugin<T extends ProjectData, U extends ProjectData> = (project: T) => U`. It'd be nice to be able to say e.g. `vite.addBuilder` has type `Plugin<{ node_modules: Package }, { viteBundle: Package }>`. (Modulo names. I don't like the name `Plugin`, any better ideas?) But toying around with that now I think that providing a type synonym that works well is not that easy. So maybe this PR should include a type synonym and have some type tests for that?

File: ts/project.test.ts [hunk 1]
```diff
@@ -50,3 +56,157 @@ const _testTypeCheckingOfAddExecutableTemplate = (project: Project) => {
   // @ts-expect-error - shell should be the actual executable now, not the helper
   p.shell``;
 };
+
+describe("Project.add", () => {
+  it("allows to add fields with .add", () => {
```
Comment (jkarni) [inline ts/project.test.ts:61]:
```suggestion
  it("allows adding fields with .add", () => {
```
here and in a few places below

**Comment by jkarni** [comment]
> Is .add the best name? Some alternatives are .with, .modify, .withSelf. Any other ideas? I kind of like both .add and .with. Although, technically, this could be used to not add something but somehow modify the project in another way, including removing things. So maybe .add is not the best name.

How about `apply`? It seems the most accurate.

File: ts/project.test.ts (hunk 1 above)
Comment (soenkehahn) [reply ts/project.test.ts]: Done in a515a1ea.

**Comment by soenkehahn** [comment]
> How about `apply`? It seems the most accurate.

We now changed the api of `.add` (and `Plugin`) so that it actually doesn't allow removing fields from a `Project`, see https://github.com/garnix-io/garn/blob/a515a1ea98f04244068b66aebda150067c40f17e/ts/project.ts#L222-L230. (Maybe it's possible by doing something weird, like setting a field to `undefined`, but it's certainly not an obvious way to use the API.) So I think the name `.add` makes a lot of sense and is consistent with the other functions that start with `add`.

**Review by soenkehahn — APPROVED** [review]

**Review by alexdavid — COMMENTED** [review]
This looks good to me too. Going to merge

**Pull request merged**